src/io.jl

```diff
@@ -45,13 +45,95 @@ DateTime. This is used to convert between the solver's inner float time, and the
 """
 time_since(t::Real, t0::DateTime)::DateTime = t0 + Millisecond(round(1000 * t))
 
-function write_basin_output(reg::Register)
-    (; config, integrator) = reg
+"""
+    load_data(db::DB, config::Config, tablename::String)::Union{Table, Query, Nothing}
+
+Load data from Arrow files if available, otherwise the GeoPackage.
+Returns either an `Arrow.Table`, `SQLite.Query` or `nothing` if the data is not present.
+"""
+function load_data(db::DB, config::Config, tablename::String)::Union{Table, Query, Nothing}
+    # TODO reverse nodetype and kind order in TOML
+    nodetype, kind = split_tablename(tablename)
+    path = getfield(getfield(config, kind), nodetype)
+    if !isnothing(path)
+        table_path = input_path(config, path)
+        return Table(read(table_path))
+    end
+
+    if exists(db, tablename)
+        return execute(db, string("select * from '$tablename'"))
+    end
+
+    return nothing
+end
+
+function load_dataframe(
+    db::DB,
+    config::Config,
+    tablename::String,
+)::Union{DataFrame, Nothing}
+    query = load_data(db, config, tablename)
+    if isnothing(query)
+        return nothing
+    end
+
+    df = DataFrame(query)
+    if hasproperty(df, :time)
+        df.time = DateTime.(df.time)
+    end
+    return df
+end
+
+function load_required_data(
+    db::DB,
+    config::Config,
+    tablename::String,
+)::Union{Table, Query, Nothing}
+    data = load_data(db, config, tablename)
+    if data === nothing
+        error("Cannot find data for '$tablename' in Arrow or GeoPackage.")
+    end
+    return data
+end
+
+"Construct a path relative to both the TOML directory and the optional `input_dir`"
+function input_path(config::Config, path::String)
+    return normpath(config.toml_dir, config.input_dir, path)
+end
+
+"Construct a path relative to both the TOML directory and the optional `output_dir`"
+function output_path(config::Config, path::String)
+    return normpath(config.toml_dir, config.output_dir, path)
+end
+
+parsefile(config_path::AbstractString) =
+    from_toml(Config, config_path; toml_dir = dirname(normpath(config_path)))
+
+# Read into memory for now with read, to avoid locking the file, since it mmaps otherwise.
+# We could pass Mmap.mmap(path) ourselves and make sure it gets closed, since Arrow.Table
+# does not have an io handle to close.
+_read_table(entry::AbstractString) = Arrow.Table(read(entry))
```

Comment:
Why prefix with an underscore?

Reply:
Sometimes it's done to signify internal methods. Perhaps it could be renamed though, since we don't use leading underscores elsewhere in the code. Although this particular function is currently not used, it will become relevant again for the validation, when that is hooked up again.

Reply:
Too bad that we have to rely on python-like private conventions instead simply not exporting functions from Julia modules :)